core/src/main/java/com/linecorp/armeria/internal/common/RequestContextUtil.java

```diff
@@ -236,7 +228,36 @@ private static AutoCloseable invokeHook(RequestContext ctx) {
             return null;
         }
 
-        return closeable;
+        return () -> {
+            try {
+                closeable.close();
+            } catch (Throwable t) {
+                logger.warn("{} Unexpected exception while closing RequestContext.hook().", ctx, t);
+            }
+        };
+    }
+
+    public static Supplier<AutoCloseable> mergeHooks(Supplier<? extends AutoCloseable> hook1,
+                                                     Supplier<? extends AutoCloseable> hook2) {
+        if (hook1 == NOOP_CONTEXT_HOOK) {
+            //noinspection unchecked
+            return (Supplier<AutoCloseable>) hook2;
+        } else if (hook2 == NOOP_CONTEXT_HOOK) {
+            //noinspection unchecked
+            return (Supplier<AutoCloseable>) hook1;
+        } else {
+            return () -> {
+                final AutoCloseable closeable1 = hook1.get();
+                final AutoCloseable closeable2 = hook2.get();
+                return () -> {
+                    try {
+                        closeable1.close();
+                    } finally {
+                        closeable2.close();
```

Comment:
I know that the original logic was implemented in this way, but shouldn't it be?
```suggestion
                        closeable2.close();
                    } finally {
                        closeable1.close();
```